### PR TITLE
[examples] SummarizationDataset cleanup

### PR DESCRIPTION
**Pull request by sshleifer**
- factor out redundant tokenization logic
-  For both articles and summaries, batches are "trimmed" such that no columns are full of `pad_token_id`. 
- The max sizes are 1024 for source and 56 for target. This ensures that truncation never happens for summaries, and rarely happens for articles. These values are unchanged, but converted to command line arguments for ease of use.
- added small unittest (just for the dataset).
- Verified manually on GPU. Loss goes down, peak memory usage identical, speed improved by ~10\%.

Summary Statistics on summary lengths (in # tokens) for cnn_dm test data:
image

### CODE REVIEW

**Comment by sshleifer** [comment]
@acarrera94 does this look OK?

**Comment by acarrera94** [comment]
This looks great! Something else we might want to figure out is the best configuration of max_seq_length and max_target_length. For example, the tesla V100 in google colab can for sure handle about max_seq_length=768 and max_target_length=56 with a batch size of 4. It can't handle max_seq_length=1028 with the same configuration, since it will run out of memory.

**Comment by sshleifer** [comment]
Going to address by trimming batches so that they don't add extra padding.

**Comment by sshleifer** [comment]
Hard to unittest cause of the PL dependency but I verified that the script runs locally.

File: examples/summarization/bart/run_bart_sum.py [hunk 1]
```diff
@@ -77,10 +71,7 @@ def test_step(self, batch, batch_idx):
             self.tokenizer.decode(g, skip_special_tokens=True, clean_up_tokenization_spaces=True)
             for g in generated_ids
         ]
-        target = [
-            self.tokenizer.decode(t, skip_special_tokens=True, clean_up_tokenization_spaces=True)
-            for t in batch["target_ids"]
-        ]
+        target = [self.tokenizer.decode(t, skip_special_tokens=True, clean_up_tokenization_spaces=True) for t in y]
```
Comment (patrickvonplaten) [inline examples/summarization/bart/run_bart_sum.py]: `for t in y` could probably get a better name

File: examples/summarization/bart/run_bart_sum.py [hunk 2]
```diff
@@ -19,30 +19,20 @@ class BartSystem(BaseTransformer):
     mode = "language-modeling"
 
     def __init__(self, hparams):
-        super(BartSystem, self).__init__(hparams, num_labels=None, mode=self.mode)
+        super().__init__(hparams, num_labels=None, mode=self.mode, output_past=False)
 
-    def forward(
-        self, input_ids, attention_mask=None, decoder_input_ids=None, decoder_attention_mask=None, lm_labels=None
-    ):
+    def forward(self, input_ids, attention_mask=None, decoder_input_ids=None, lm_labels=None):
```
Comment (patrickvonplaten) [inline examples/summarization/bart/run_bart_sum.py]: Maybe remove the `forward()` function - it just transfers the values to `self.model.forward` no?

**Comment by srush** [comment]
@sshleifer, in MT code it is common to group the batches together to minimize padding. Do you think this is worth implementing? One downside of this method is that is doesn't really work on TPU or architectures that expect fixed sizes.

File: examples/summarization/bart/run_bart_sum.py (hunk 2 above)
Comment (srush) [reply examples/summarization/bart/run_bart_sum.py]: You need this for lightning.

**Comment by sshleifer** [comment]
Yes, is something like [SortishSampler](https://github.com/fastai/fastai/blob/master/fastai/text/data.py#L99) the right idea?

It seems easy to implement if you only consider padding on the `source` side. Do you know of an intelligent `key_func` to sort examples that considers both sides?

File: examples/summarization/bart/run_bart_sum.py [hunk 3]
```diff
@@ -101,11 +92,21 @@ def test_epoch_end(self, outputs):
...
+            max_target_length=self.hparams.max_target_length,
         )
-        dataloader = DataLoader(train_dataset, batch_size=self.hparams.train_batch_size)
+
+    def get_dataloader(self, type_path: str, batch_size: int) -> DataLoader:
+        dataset = SummarizationDataset(self.tokenizer, type_path=type_path, **self.dataset_kwargs)
+        dataloader = DataLoader(dataset, batch_size=batch_size, collate_fn=dataset.collate_fn)
+        return dataloader
+
+    def train_dataloader(self) -> DataLoader:
```
Comment (LysandreJik) [inline examples/summarization/bart/run_bart_sum.py:108]: Love the typings

File: examples/summarization/bart/test_bart_examples.py [hunk 4]
```diff
@@ -6,27 +6,58 @@
...
         testargs = ["evaluate_cnn.py", str(tmp), output_file_name]
         with patch.object(sys, "argv", testargs):
             _run_generate()
             self.assertTrue(Path(output_file_name).exists())
             os.remove(Path(output_file_name))
+
+    def test_bart_summarization_dataset(self):
+        tmp_dir = Path(tempfile.gettempdir())
+        articles = [" Sam ate lunch today", "Sams lunch ingredients"]
+        summaries = ["A very interesting story about what I ate for lunch.", "Avocado, celery, turkey, coffee"]
```
Comment (LysandreJik) [inline examples/summarization/bart/test_bart_examples.py:44]: Nice!

**Review by LysandreJik — APPROVED** [review]
LGTM

**Pull request merged**